### ISSUE TITLE
TODO/DRAFT todo states vs. HUGO_DRAFT property

### ISSUE THREAD

**Issue by akirak** [kaushalmodi/ox-hugo#171]
Hello, thanks for commenting on [my blog post](https://jingsi.space/post/2018/03/13/simplified-configuration-for-hugo/).

The current spec of ox-hugo sets the draft status if the TODO keyword is either `TODO` or `DRAFT`, but I have a much more complex setup of TODO keywords in org-mode. I personally prefer marking a post as final if and only if it has explicit `DONE` state, and otherwise it should be considered as a draft.

To implement this behavior, I set `HUGO_DRAFT` property explicitly before dispatching ox-hugo using a wrapper function:

- https://github.com/akirak/emacs.d/blob/53475e6be7df2ee30c799586872f145183fdd040/org/init-org-hugo.el#L57
- https://github.com/akirak/emacs.d/blob/53475e6be7df2ee30c799586872f145183fdd040/org/init-org-hugo.el#L103

However, it seems that the draft status from the TODO state takes precedence over explicit draft property: https://github.com/kaushalmodi/ox-hugo/blob/master/ox-hugo.el#L2777

I think this is a strange behavior, because something made explicit should always take precedence over implicitness. Actually, my workaround still works this way, because if the todo state is neither `TODO` nor `DRAFT`, the explicit draft status is considered by ox-hugo. 

What do you think? Is there any reason for `HUGO_DRAFT` not taking precedence over `TODO`/`DRAFT` states?

**Comment by kaushalmodi** [comment]
Thanks for opening this issue. 

> The current spec of ox-hugo sets the draft status if the TODO keyword is either TODO or DRAFT, but I have a much more complex setup of TODO keywords in org-mode. I personally prefer marking a post as final if and only if it has explicit DONE state, and otherwise it should be considered as a draft.

Would it work if in the draft status detection `cond` form, a "DONE" case is added too, the the default draft state is made configurable? At present, the default draft state is "false". I can make it user-configurable via a defcustom, so that you can set it to "true".

Would those 2 changes be enough to make the draft state set exact as you want?

> To implement this behavior, I set HUGO_DRAFT property explicitly before dispatching ox-hugo using a wrapper function:

>     https://github.com/akirak/emacs.d/blob/53475e6be7df2ee30c799586872f145183fdd040/org/init-org-hugo.el#L57
>     https://github.com/akirak/emacs.d/blob/53475e6be7df2ee30c799586872f145183fdd040/org/init-org-hugo.el#L103

You seem to have a great deal of Hacking. But hopefully the above suggested fix removed the need to do all that. 

> However, it seems that the draft status from the TODO state takes precedence over explicit draft property: https://github.com/kaushalmodi/ox-hugo/blob/master/ox-hugo.el#L2777

> I think this is a strange behavior, because something made explicit should always take precedence over implicitness. Actually, my workaround still works this way, because if the todo state is neither TODO nor DRAFT, the explicit draft status is considered by ox-hugo.

> What do you think? Is there any reason for HUGO_DRAFT not taking precedence over TODO/DRAFT states?

I see the setting of TODO/DONE/DRAFT states vs  setting the HUGO_DRAFT property as equals i.e. both can be explicitly used to set the draft state.  So I had to give one the precedence over the other.

Setting the TODO/DONE states in headings is more convenient than setting the Org property EXPORT_HUGO_DRAFT. Also the draft state of a subtree is more apparent when using those TODO states. So I prefer to have those at a higher precedence.  

The `EXPORT_HUGO_DRAFT` is there just to support the per-file export flow. 

*I'm reading/replying to this from a phone. Let me know if I misunderstood your requirement in the above proposed solution. I'll commit the proposed fix on Monday.*

**Comment by kaushalmodi** [comment]
> I set the timestamp of the post explicitly based on the draft status. If the post is a draft, EXPORT_DATE is set to the current time so that the (blog) post is displayed on the top,

I assume you set that date just so that the blog post is shown at the top while running `hugo server`? 

*You might already know, but running `hugo server` with `-D --navigateToChanged` will take you to the last updated draft post regardless of whether its date is set.*

> but otherwise EXPORT_DATE is deleted, so the published date with be the same as its CLOSED date. 

As I said above, you don't need to add that EXPORT_DATE if you are doing so just to easily access the draft page from the Hugo list page. But if you still need it for a different reason, then OK. Even then, you don't then need to remove it, because the TODO state based time stamp has a higher precedence than the `EXPORT_DATE` set date (consistency). 

> I have to retrieve the draft status in my script, so I will still need the hack. This is crazy, but I couldn't come up with a simpler solution.

If you still need to do this, **would it help if I break out that draft state parsing to a separate function?** Then you can advice that, and hopefully reduce the hack code? 

I see that you are using functions with `org-hugo--*` (with double hyphens i.e. internal functions). I wouldn't recommend using those as they are strictly for internal use (i.e. not for use in user config). So they might disappear/change signature/behavior/etc without a clear notice. 

Feel free to open a separate issue to explain the need for using those tags/categories based workaround in your config.

**Comment by kaushalmodi** [comment]
Also, can you paste the snippets for the following:

1. Org snippet of your draft state post
2. Expected Markdown front-matter (draft, date values)
3. Org snippet of the same post converted to DONE.
4. Expected Markdown front-matter (draft, date values) after that DONE conversion. 

I am trying to see if we can do without any of your ox-hugo hack even with your custom TODO state flow.

**Comment by kaushalmodi** [comment]
> I know what you mean. I'm only using org-hugo--get-tags, and it's an alias to org-get-tags, so I probably should use it directly.

That also might not be a good idea as that function behavior changes between Org 9.1 and 9.2 (development version). So, if you use the unaliased version, your setup will break when switching between Org 9.1 and 9.2.

**Comment by akirak** [comment]
I see, thanks!

**Comment by kaushalmodi** [comment]
Hello, it's not clear why this issue was closed. 

Do you need to have any changes made in ox-hugo so that you can avoid using workarounds in your config?

Would my earlier suggestion work? You probably missed reading that part. .. and also my comment in https://github.com/kaushalmodi/ox-hugo/issues/171#issuecomment-403287988 requesting minimum workable examples from you. 

> Would it work if in the draft status detection cond form, a "DONE" case is added too, the the default draft state is made configurable? At present, the default draft state is "false". I can make it user-configurable via a defcustom, so that you can set it to "true".

> Would those 2 changes be enough to make the draft state set exact as you want?

**Comment by akirak** [comment]
I closed this issue because I thought I finally understood your point. Your design makes sense and probably don't need the change. 

> Would it work if in the draft status detection cond form, a "DONE" case is added too, the the default draft state is made configurable? At present, the default draft state is "false". I can make it user-configurable via a defcustom, so that you can set it to "true".

> Do you need to have any changes made in ox-hugo so that you can avoid using workarounds in your config?

I don't mind whether you do it or not. It would be better if you make those changes, but even if you don't, I can manage it. My goal is not to avoid any hacks but to get the desired results from your package. And I already do. If there is a breaking change in your package, I will just catch up with it.

Yes, the instruction in the comment seemed to work. Thanks for the idea.

**Comment by kaushalmodi** [comment]
@akirak Check out the changes and the new tests added in https://github.com/kaushalmodi/ox-hugo/commit/76109cd97c71924fb725302d2fa18f469e905db8. For your use-case, you can now set `org-hugo-default-draft-state` to `t`.

**Comment by kaushalmodi** [comment]
@akirak OK, see https://github.com/kaushalmodi/ox-hugo/commit/8ab749c640985d24c47616cd3e44cd865a84be45 instead.. that should work.

**Comment by kaushalmodi** [comment]
@akirak I am just being too hasty.. the latest commit is still incorrect.. Let me fix and test things.. I'll update you when all things are green.

**Comment by kaushalmodi** [comment]
All good now :) Please update to the latest version of ox-hugo.

- You can close this issue if the update does things as we discussed.
- Let me know if this helps reduce some of your custom code.
- Also, feel free to open a separate issue if you'd like to discuss your custom code for ox-hugo tags/categories.

**Comment by kaushalmodi** [comment]
In a later commit, I also add a new defcustom `org-hugo-todo-keywords` which you might like to customize to add the NEXT state: https://github.com/kaushalmodi/ox-hugo/commit/85b1377948202e612a8f130befa13c125de0fe6d

**Comment by akirak** [comment]
I'd prefer "negating" the default draft status for specific TODO keywords rather than enumerating all possible TODO keywords in `org-hugo-todo-keywords`, but this view may be opinionated. What do you think? I don't understand why you have to distinguish between `todo` and `draft` states.

**Comment by kaushalmodi** [comment]
> I'm not sure if org-hugo-todo-keywords is a good idea. 

I agree.  I'll remove this defcustom. I have a better idea with which you wouldn't need to do any extra config for ox-hugo. 

> I appreciate your effort to develop this package.

Thanks. 

This particular change was mainly to remove the hard-coding of TODO states that I had earlier. I thought that adding a defcustom would allow people to use any Org TODO keyword they like. But I can see how adding that user-error is an inconvenience. 

About the todo vs draft states, yes, that's a bit opinionated... The "TODO" state is something that I might work on a blog post some day.. I have about a dozen of such posts. So when I do "C-c C-e H A", I don't want to have a dozen echoes telling that those posts are drafts.. I already know that. 

The "DRAFT" state is for the one or two posts that I am actively working on, but not yet ready to publish. For such, having that echo is useful (to me). 

But it's understood that having a special "DRAFT" state wouldn't fit in everyone's flow. 

I'll remove this defcustom, and I have a better idea to remove the hard-coding of TODO and DONE states (`org-done-keywords`).. only the DRAFT state will be hard-coded for now.. and no errors.

**Comment by akirak** [comment]
Thank you for your concern. I hope ox-hugo will be even better!

**Comment by kaushalmodi** [comment]
Can I get a quick feedback: Do you even need the default draft defcustom? If I understand correctly, you probably have all your post subtrees marked with *some* TODO state.. the default draft state affects only the subrees without any TODO keyword.

**Comment by kaushalmodi** [comment]
Hmm, would it be convenient to have a TODO/NEXT state for all the subtrees that you don't intend to publish (draft = true)? Then I can default draft = false for all post subtrees without a TODO keyword.. I'm just brainstorming if I even need that default draft defcustom..

Another way to draft a batch of posts is to nest them in a subtree with EXPORT_HUGO_DRAFT property set to true. All the nested posts will then be drafts by property inheritance (ox-hugo enforces that inheritance). Then gradually as a new post becomes ready to publish, you set the DONE state for todo (DONE todo state overrides that property).

**Comment by kaushalmodi** [comment]
.. or just have `#+hugo_draft: true` at the top of the Org file.. that will cause all post subtrees **without DONE state** to be drafts by default.

Thanks for this discussion.. looks like that default draft defcustom isn't needed after all.. let me know if you think otherwise.

**Comment by kaushalmodi** [comment]
I removed those 2 defcustoms in the last commit to ox-hugo. Thanks for this nudge! Using `org-done-keywords` made the code much simpler, while getting the same results and lesser maintenance overhead (syncing the keywords between `org-todo-keywords` and now removed `org-hugo-todo-keywords`).

I also added few test cases showing use of keywords that are not `TODO` or `DONE` [here][1].

Based on my [above comment][2], there shouldn't be any need for `org-hugo-default-draft-state` either.. let me know if you find that not to be the case.

[1]: https://github.com/kaushalmodi/ox-hugo/commit/41710e542ad2062b4dac346a2535f04115d0dd82#diff-6fa190cf16cc1fe9ba0fa6673ddb5970R6302
[2]: https://github.com/kaushalmodi/ox-hugo/issues/171#issuecomment-403792998
